src/test/java/org/apache/sysds/test/functions/builtin/BuiltinImageRotateTest.java

```diff
@@ -0,0 +1,102 @@
+		final double fill_value = 128.0;
+		double[][] input = TestUtils.readMatrixFromFile(this.getClass().getResource("ImageTransformInput"), h, w);
+		double[][] reference = input;
+		String HOME = SCRIPT_DIR + TEST_DIR;
+		fullDMLScriptName = HOME + TEST_NAME + ".dml";
+		programArgs = new String[] {"-nvargs", "in_file=" + input("A"), "out_file=" + output("B"), "width=" + cols,
+				"height=" + rows, "angle=" + 0, "fill_value=" + fill_value};
+		writeInputMatrixWithMTD("A", input, true);
+		runTest(true, false, null, -1);
+
+		HashMap<MatrixValue.CellIndex, Double> dmlfile = readDMLMatrixFromOutputDir("B");
+		double[][] dml_res = TestUtils.convertMatrix(dmlfile, h, w);
+		TestUtils.compareMatrices(reference, dml_res, eps, "Input vs. DML");
+	}
+
+	@Test public void testImageRotateComplete() throws Exception {
+		loadTestConfiguration(getTestConfiguration(TEST_NAME));
+		final int w = 500, h = 135;
+		final double fill_value = 128.0;
+		double[][] input = TestUtils.readMatrixFromFile(this.getClass().getResource("ImageTransformInput"), h, w);
+		double[][] reference = input;
+		String HOME = SCRIPT_DIR + TEST_DIR;
+		fullDMLScriptName = HOME + TEST_NAME + ".dml";
+		programArgs = new String[] {"-nvargs", "in_file=" + input("A"), "out_file=" + output("B"), "width=" + cols,
+				"height=" + rows, "angle=" + (2 * Math.PI), "fill_value=" + fill_value};
+		writeInputMatrixWithMTD("A", input, true);
+		runTest(true, false, null, -1);
+
+		HashMap<MatrixValue.CellIndex, Double> dmlfile = readDMLMatrixFromOutputDir("B");
+		double[][] dml_res = TestUtils.convertMatrix(dmlfile, h, w);
+		TestUtils.compareMatrices(reference, dml_res, eps, "Input vs. DML");
+	}
+
+	@Test public void testImageRotatePillow() throws Exception {
+		loadTestConfiguration(getTestConfiguration(TEST_NAME));
+		final int w = 500, h = 135;
+		final double fill_value = 128.0;
+		double[][] input = TestUtils.readMatrixFromFile(this.getClass().getResource("ImageTransformInput"), h, w);
+		double[][] reference = TestUtils.readMatrixFromFile(this.getClass().getResource("ImageTransformRotated"), h, w);
+		String HOME = SCRIPT_DIR + TEST_DIR;
+		fullDMLScriptName = HOME + TEST_NAME + ".dml";
+		programArgs = new String[] {"-nvargs", "in_file=" + input("A"), "out_file=" + output("B"), "width=" + cols,
+				"height=" + rows, "angle=" + angle, "fill_value=" + fill_value};
+		writeInputMatrixWithMTD("A", input, true);
+		runTest(true, false, null, -1);
+
+		HashMap<MatrixValue.CellIndex, Double> dmlfile = readDMLMatrixFromOutputDir("B");
+		double[][] dml_res = TestUtils.convertMatrix(dmlfile, h, w);
+		TestUtils.compareMatrices(reference, dml_res, eps, "Pillow vs. DML");
+		// The error here seems to be in the pillow implementation. It calculates that the source coordinates of
+		// (339, 14) are (351, 87) using fixed point arithmetic, while the floating point implementation returns
+		// (351.975384 88.000514) rounded down to (351, 88). (All indices here start at 0)
```

Comment:
this difference sounds large, could it be possible to test with some other framework, or write the calculation in java?

Reply:
I analyzed the source code of the Pillow library. They use a fixed-point arithmetic implementation by default which for the given input returns 87 as y coordinate. If the given input does not satisfy some criteria they fall back on a floating-point arithmetic implementation. For the given input the floating-point implementation returns 88 as y coordinate which is the same as my floating-point implementation in dml.